### ISSUE TITLE
[Yaml] ~ is not interpreted as null when used as array key

### ISSUE THREAD

**Issue by ivoba** [symfony/symfony#23664]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.3.5

When using ~ as array key its interpreted as string "~" since i upgraded to symfony 3.3.
Before it was interpreted as null resp. "".

As it is in Yaml spec it should be interpreted as null even as key, i suppose:
http://www.yaml.org/type/null.html

Example:

```
my_service:
  class: SomeClass
  arguments:
    -
     ~: 'default'
     key: 'value'
```
Arguments results in:

> array:2 [▼
>   "~" => "default"
>   "key" => "value"
> ]

Expected would have been:

> array:2 [▼
>   "" => "default"
>   "key" => "value"
> ]

Is this intended behaviour?

**Comment by javiereguiluz** [comment]
I'd say this is a bug, but our main Yaml maintainer @xabbuh will confirm. Thanks for reporting!

**Comment by xabbuh** [comment]
Well, inside the `YamlFileLoader` of the DependencyInjection component we use the `PARSE_KEYS_AS_STRING` flag which evaluates `~` as `'~'`. So IMO that's expected.

**Comment by xabbuh** [comment]
In 3.4 you will additionally see a message that advices you to quote evaluable keys.

**Comment by ivoba** [comment]
Ok, thanks for clarification.

**Comment by ivoba** [comment]
@xabbuh shouldnt then the quoted tilde be interpreted as null? Or will this come in 3.4?
```
my_service:
  class: SomeClass
  arguments:
    -
     "~": 'default'
     key: 'value'
```
Currently its not.

**Comment by javiereguiluz** [comment]
@ivoba I'd say that quoted `'~'` is just a string, so the current behavior is OK. However, unquoted `~` is always null in Yaml, so I still think this is a bug. (Not very important because it's strange to use null as the key value).

**Comment by GuilhemN** [comment]
This will be fixed in 4.0: keys will always be evaluated and `~` will be forbidden as key (as it's forbidden in php).

**Comment by javiereguiluz** [comment]
@GuilhemN but `null` keys work in PHP, right? See https://3v4l.org/mLn6T

**Comment by xabbuh** [comment]
The actual key is the empty string (PHP casts `null` to a string).

**Comment by GuilhemN** [comment]
This can be closed imo, it is already fixed in 4.0.
